### PR TITLE
fix: parse deep link correctly

### PR DESCRIPTION
**Pull request by Antasel**
### Fixed Issues
$ https://github.com/Expensify/App/issues/16762
Proposal: https://github.com/Expensify/App/issues/16762#issuecomment-1682687000

### Tests

 Go to a chat and send following test strings and verify that urls and emails are parsed as expected.
<details>
  <summary>Test Strings</summary>

```
test@expensify.com https://www.expensify.com
test@expensify.com-https://www.expensify.com
test@expensify.com/https://www.expensify.com
test@expensify.com?https://www.expensify.com
test@expensify.com>https://www.expensify.com
https://staging.new.expensify.com/details/test@expensify.com
staging.new.expensify.com/details
https://www.expensify.com?name=test&email=test@expensify.com
https://staging.new.expensify.com/details?login=testing@gmail.com
staging.new.expensify.com/details?login=testing@gmail.com
http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled
-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com
expensify.com -expensify.com @expensify.com
https//www.expensify.com
//www.expensify.com?name=test&email=test@expensify.com
//staging.new.expensify.com/details?login=testing@gmail.com
/details?login=testing@gmail.com
?name=test&email=test@expensify.com
example.com/https://www.expensify.com
test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com
test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com
test@gmail.com-https://staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com

test@gmail.com/https://example.com/google@email.com?email=asd@email.com
test@gmail.com/test@gmail.com/https://example.com/google@email.com?email=asd@email.com
```
</details>

### Demo Video:

[Screencast from 6-9-23 09:04:20.webm](https://github.com/Expensify/expensify-common/assets/139204153/28b64dc8-de47-45e2-9268-679e7b58d926)

### Autotest cases:
<details>
  <summary>ExpensiMark-HTML-test.js new tests</summary>

```
test('Test urls autolinks correctly', () => {
    let testString = 'test@expensify.com https://www.expensify.com\n' +
    'test@expensify.com-https://www.expensify.com\n' +
    'test@expensify.com/https://www.expensify.com\n' +
    'test@expensify.com?https://www.expensify.com\n' +
    'test@expensify.com>https://www.expensify.com\n' +
    'https://staging.new.expensify.com/details/test@expensify.com\n' +
    'staging.new.expensify.com/details\n\n' +
    'https://www.expensify.com?name=test&email=test@expensify.com\n' +
    'https://staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'staging.new.expensify.com/details?login=testing@gmail.com\n' +
    'http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled\n' +
    '-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com\n' +
    'expensify.com -expensify.com @expensify.com\n' +
    'https//www.expensify.com\n' +
    '//www.expensify.com?name=test&email=test@expensify.com\n' +
    '//staging.new.expensify.com/details?login=testing@gmail.com\n' +
    '/details?login=testing@gmail.com\n' +
    '?name=test&email=test@expensify.com\n\n' +
    'example.com/https://www.expensify.com\n' +
    'test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com-https://staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com\n' +
    'test@gmail.com/https://example.com/google@email.com?email=asd@email.com\n' +
    'test@gmail.com/test@gmail.com/https://example.com/google@email.com?email=asd@email.com';

    let resultString = '<a href="mailto:test@expensify.com">test@expensify.com</a> <a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>/<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="mailto:test@expensify.com">test@expensify.com</a>?<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' + 
    '<a href="mailto:test@expensify.com">test@expensify.com</a>&gt;<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details/test@expensify.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details/test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details</a><br /><br />'+ 
    '<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '<a href="http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled" target="_blank" rel="noreferrer noopener">http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled</a><br />' +
    '-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> /<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> @https://www.expensify.com<br />' +
    '<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> -<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> @expensify.com<br />' +
    'https//<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com</a><br />' +
    '//<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com?name=test&amp;email=test@expensify.com</a><br />' +
    '//<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a><br />' +
    '/details?login=<a href="mailto:testing@gmail.com">testing@gmail.com</a><br />' +
    '?name=test&amp;email=<a href="mailto:test@expensify.com">test@expensify.com</a><br /><br />' +
    '<a href="https://example.com/https://www.expensify.com" target="_blank" rel="noreferrer noopener">example.com/https://www.expensify.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> <a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a> //<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a>-<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a><br />' +
    '<a href="mailto:test@gmail.com\">test@gmail.com</a>/<a href="https://example.com/google@email.com?email=asd@email.com" target="_blank" rel="noreferrer noopener">https://example.com/google@email.com?email=asd@email.com</a><br />' +
    '<a href="mailto:test@gmail.com">test@gmail.com</a>/<a href="mailto:test@gmail.com">test@gmail.com</a>/<a href="https://example.com/google@email.com?email=asd@email.com" target="_blank" rel="noreferrer noopener">https://example.com/google@email.com?email=asd@email.com</a>';
    expect(parser.replace(testString)).toBe(resultString);
});
```
</details>

### CODE REVIEW

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -476,9 +470,28 @@ export default class ExpensiMark {
             }
             replacedText = replacedText.concat(textToCheck.substr(startIndex, (match.index - startIndex)));
 
+            // we want to avoid matching email address domains
```
Comment (cead22) [inline lib/ExpensiMark.js:473]:
```suggestion
            // We want to avoid matching email address domains
```
This comment should also say "why"

File: lib/ExpensiMark.js [hunk 2]
```diff
@@ -476,9 +470,28 @@ export default class ExpensiMark {
...
+            let abort = false;
+            let shouldRetryByAtSign = false;
```
Comment (cead22) [inline lib/ExpensiMark.js:475]: What does "by at sign" mean? are we doing another try by removing the @ sign, by adding the @ sign, or by testing against a new regex with/out @ sign?

File: lib/ExpensiMark.js [hunk 3]
```diff
@@ -476,9 +470,28 @@ export default class ExpensiMark {
...
+
+            if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
+                const domainRegex = new RegExp('^(([a-z-0-9]+\\.)+[a-z]{2,})(\\S*)', 'i');
+                const domainMatch = domainRegex.exec(match[2]);
+
+                // e.g. test@expensify.com/https://www.test.com
+                // If the matched string faces @ sign before it,
+                // We will retry to apply autolink rule to the string(e.g. /https://www.test.com) except for domain(e.g. expensify.com) after @ sign.
```
Comment (cead22) [inline lib/ExpensiMark.js:483]: Can you update this comment to make it clearer?
- What is the e.g. an example of? shouldn't that be after another comment?
- What does "faces @ sign before it"? I suggest "If the matches string has a leading @ sign" -- use "leading" or "trailing" depending on the case
- It should say "why"

File: lib/ExpensiMark.js [hunk 4]
```diff
@@ -476,9 +470,28 @@ export default class ExpensiMark {
...
+                if ((domainMatch !== null) && (domainMatch[3] !== '')) {
+                    shouldRetryByAtSign = true;
+                    replacedText = replacedText.concat(domainMatch[1] + this.replace(domainMatch[3], {filterRules: ['autolink']}));
+                } else {
+                    abort = true;
```
Comment (cead22) [inline lib/ExpensiMark.js:488]: Would it make sense to add a comment explaining why we're setting abort to true here? It's not immediately obvious to me

File: __tests__/ExpensiMark-HTML-test.js [hunk 5]
```diff
@@ -609,6 +609,60 @@ test('Test a url ending with a closing parentheses autolinks correctly', () => {
     expect(parser.replace(testString)).toBe(resultString);
 });
 
+test('Test urls autolinks correctly', () => {
+    let testString = 'test@expensify.com https://www.expensify.com\n' +
```
Comment (cead22) [inline __tests__/ExpensiMark-HTML-test.js:613]: This is a little hard to read and review, so I was thinking maybe we should add a test for each of these cases, but I think a better idea would be to make this an array with entries for each case, and I think even better would be an array of arrays, where the test string and the expected string are next to each other for easy comparison, readability, review, and updates.

If you agree, can you update accordingly?

**Review by cead22 — CHANGES_REQUESTED** [review]

File: lib/ExpensiMark.js (hunk 1 above)
Comment (Antasel) [reply lib/ExpensiMark.js:473]: it's from old PR, I just moved it to other line. but new comment is more readable.

File: lib/ExpensiMark.js (hunk 3 above)
Comment (Antasel) [reply lib/ExpensiMark.js:483]: > What is the e.g. an example of? shouldn't that be after another comment?

yes, it's example of comment. It should be after comment normally, but I tended to point '/https://www.test.com' and 'test@expensify.com' in comment details

> What does "faces @ sign before it"? I suggest "If the matches string has a leading @ sign" -- use "leading" or "trailing" depending on the case

Your suggestion is good, I will change comment as you recommended. "if the matched string has a leading @ sign"

> It should say "why"

By a Old PR, if matched string has leading @ sign, It simply aborts to apply autolink rule. But we have edge cases like test@expensify.com/https://www.test.com, where https://www.test.com should be parsed as url.
To explain above reason, I added a example of edge test string

File: lib/ExpensiMark.js (hunk 4 above)
Comment (Antasel) [reply lib/ExpensiMark.js:488]: image

This comment is to say why we are setting abort to true

File: __tests__/ExpensiMark-HTML-test.js (hunk 5 above)
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:613]: How about dividing it into 3~4 test sections according to test string type, like https://github.com/Expensify/App/issues/16762#issuecomment-1598810396

Comment (cead22) [reply __tests__/ExpensiMark-HTML-test.js:613]: That works, but I still arrays are better than string concats, and an array of array such that the test and expected strings are next to each other is better than both

File: lib/ExpensiMark.js (hunk 2 above)
Comment (Antasel) [reply lib/ExpensiMark.js:475]: I am also feeling that variable name is not reasonable.

I think `isReparsedByAutolink` is more readable

File: __tests__/ExpensiMark-HTML-test.js (hunk 5 above)
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:613]: The following implementation is acceptable ? 

```
    const testCases = [
        {
            testString: 'test@expensify.com https://www.expensify.com',
            resultString: '<a href="mailto:test@expensify.com">test@expensify.com</a> <a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a>',
        },
        {
            testString: 'test@expensify.com-https://www.expensify.com',
            resultString: '<a href="mailto:test@expensify.com">test@expensify.com</a>-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a>',
        },
    ];

    testCases.forEach(testCase => {
        expect(parser.replace(testCase.testString)).toBe(testCase.resultString);
    });
```

Comment (cead22) [reply __tests__/ExpensiMark-HTML-test.js:613]: Yes that looks better than what I had in mind :)

File: lib/ExpensiMark.js (hunk 2 above)
Comment (Antasel) [reply lib/ExpensiMark.js:475]: How do you think about new one?

**Comment by Antasel** [comment]
@cead22 Could you please check new commit https://github.com/Expensify/expensify-common/pull/573/commits/8bc8591c1fed41a6452fe57267a245cbe9654129?

File: lib/ExpensiMark.js [hunk 6]
```diff
@@ -476,9 +470,29 @@ export default class ExpensiMark {
...
+                if ((domainMatch !== null) && (domainMatch[3] !== '')) {
+                    isReparsedByAutolink = true;
+                    replacedText = replacedText.concat(domainMatch[1] + this.replace(domainMatch[3], {filterRules: ['autolink']}));
+                } else {
+                    abort = true;
+                }
+            }
+
             if (abort || match[1].includes('<pre>')) {
                 replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
```
Comment (cead22) [inline lib/ExpensiMark.js:494]: I think this is still hard to understand, so here's a suggestion to make things clearer, but I'm not sure if what I'm saying in the comments, or if the updated variable names are accurate, so I could use your help
```js
            // We want to avoid matching domains in email addresses so we don't render them as URLs,
            // but we need to check if there are valid URLs after the email address and render them accordingly,
            // e.g. test@expensify.com/https://www.test.com
            let isDoneMatching = false;
            let shouldApplyAutoLinkAgain = true;

            // If we find a URL with a leading @ sign, we need look for other domains in the rest of the string
            if ((match.index !== 0) && (textToCheck[match.index - 1] === '@')) {
                const domainRegex = new RegExp('^(([a-z-0-9]+\\.)+[a-z]{2,})(\\S*)', 'i');
                const domainMatch = domainRegex.exec(match[2]);

                // If we find another domain, we apply the auto link rule again and set a flag to avoid doing it
                if ((domainMatch !== null) && (domainMatch[3] !== '')) {
                    replacedText = replacedText.concat(domainMatch[1] + this.replace(domainMatch[3], {filterRules: ['autolink']}));
                    shouldApplyAutoLinkAgain = false;
                } else {
                    // Otherwise, we're done applying rules
                    isDoneMatching = true;
                }
            }

            // Here we should probably explain in words what the code is doing, why, and what the condition means
            if (isDoneMatching || match[1].includes('<pre>')) {
                replacedText = replacedText.concat(textToCheck.substr(match.index, (match[0].length)));
            } else if (shouldApplyAutoLinkAgain) {
```

**Review by cead22 — CHANGES_REQUESTED** [review]

File: __tests__/ExpensiMark-HTML-test.js [hunk 7]
```diff
@@ -609,6 +609,111 @@ test('Test a url ending with a closing parentheses autolinks correctly', () => {
...
+            resultString: '<a href="https://example.com/https://www.expensify.com" target="_blank" rel="noreferrer noopener">example.com/https://www.expensify.com</a>',
+        },
+        {
+            testString: 'test@gmail.com staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com',
+            resultString: '<a href="mailto:test@gmail.com">test@gmail.com</a> <a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a>',
+        },
+        {
+            testString: 'test@gmail.com //staging.new.expensify.com/details?login=testing@gmail.com&redirectUrl=https://google.com',
+            resultString: '<a href="mailto:test@gmail.com">test@gmail.com</a> //<a href="https://staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com&amp;redirectUrl=https://google.com</a>',
```
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:696]: Screenshot_1

On discord, have explicit @ signs in Urls,
and on this platform as well

File: __tests__/ExpensiMark-HTML-test.js [hunk 8]
```diff
@@ -609,6 +609,111 @@ test('Test a url ending with a closing parentheses autolinks correctly', () => {
...
+            testString: 'https://staging.new.expensify.com/details/test@expensify.com',
+            resultString: '<a href="https://staging.new.expensify.com/details/test@expensify.com" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/details/test@expensify.com</a>',
+        },
+        {
+            testString: 'staging.new.expensify.com/details',
+            resultString: '<a href="https://staging.new.expensify.com/details" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details</a>',
+        },
+        {
+            testString: 'https://www.expensify.com?name=test&email=test@expensify.com',
+            resultString: '<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com?name=test&amp;email=test@expensify.com</a>',
```
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:644]: I'ts from old PR, and we have other test cases to write it in html attribute. 
I think &amp: should be present there.
https://stackoverflow.com/questions/3705591/do-i-encode-ampersands-in-a-href

File: __tests__/ExpensiMark-HTML-test.js [hunk 9]
```diff
@@ -609,6 +609,111 @@ test('Test a url ending with a closing parentheses autolinks correctly', () => {
...
+        {
+            testString: 'https//www.expensify.com',
+            resultString: 'https//<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com</a>',
+        },
+        {
+            testString: '//www.expensify.com?name=test&email=test@expensify.com',
+            resultString: '//<a href="https://www.expensify.com?name=test&amp;email=test@expensify.com" target="_blank" rel="noreferrer noopener">www.expensify.com?name=test&amp;email=test@expensify.com</a>',
+        },
+        {
+            testString: '//staging.new.expensify.com/details?login=testing@gmail.com',
```
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:675]: we should render `://staging.new.expensify.com/details?login=testing@gmail.com` as link ? 
`://` will not be rendered. it's defined from old PR. I think we can use current test case as it is

File: __tests__/ExpensiMark-HTML-test.js [hunk 10]
```diff
@@ -609,6 +609,111 @@ test('Test a url ending with a closing parentheses autolinks correctly', () => {
...
+            testString: 'http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled',
+            resultString: '<a href="http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled" target="_blank" rel="noreferrer noopener">http://necolas.github.io/react-native-web/docs/?path=/docs/components-pressable--disabled</a>',
+        },
+        {
+            testString: '-https://www.expensify.com /https://www.expensify.com @https://www.expensify.com',
+            resultString: '-<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> /<a href="https://www.expensify.com" target="_blank" rel="noreferrer noopener">https://www.expensify.com</a> @https://www.expensify.com',
+        },
+        {
+            testString: 'expensify.com -expensify.com @expensify.com',
+            resultString: '<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> -<a href="https://expensify.com" target="_blank" rel="noreferrer noopener">expensify.com</a> @expensify.com',
```
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:664]: I can't check `https://` in the test case, you were about to point other cases ?

File: __tests__/ExpensiMark-HTML-test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:675]: You mean that we need the following test case instead of with `//` ?
```
{
            testString: '://staging.new.expensify.com/details?login=testing@gmail.com',
            resultString: '://<a href="https://staging.new.expensify.com/details?login=testing@gmail.com" target="_blank" rel="noreferrer noopener">staging.new.expensify.com/details?login=testing@gmail.com</a>',
        },
```

File: lib/ExpensiMark.js (hunk 6 above)
Comment (Antasel) [reply lib/ExpensiMark.js:494]: > // Here we should probably explain in words what the code is doing, why, and what the condition means

it's out of scope for this PR.

> // If we find a URL with a leading @ sign, we need look for other domains in the rest of the string

If we find a URL with a leading @ sign, we need look for other valid URL in the rest of the string.

> // If we find another domain, we apply the auto link rule again and set a flag to avoid doing it

If we find another string trailing domain, we apply the auto link rule again and set a flag to avoid doing it

I amended something on your commenst, I think it's more accurate to explain

File: __tests__/ExpensiMark-HTML-test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (cead22) [reply __tests__/ExpensiMark-HTML-test.js:644]: Nice thanks for sharing that! TIL you can and should html-encode & signs in html attributes

File: __tests__/ExpensiMark-HTML-test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (cead22) [reply __tests__/ExpensiMark-HTML-test.js:675]: Maybe this is my old school thinking, but on browsers, a link like `://expensify.com` translates into `https://expensify.com` if the protocol being used on that page is https, and to `http://expensify.com` is the protocol is http. That's what made me think we should parse `://expensify.com` as a URL, but maybe we don't need to do that, just wanted to get more thoughts.

This isn't a blocker

File: __tests__/ExpensiMark-HTML-test.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (cead22) [reply __tests__/ExpensiMark-HTML-test.js:664]: Not sure what you mean. What I was wondering was wether the result string should be `<a href="https://expensify.com/"...` or `<a href="://expensify.com/"...` which would use whatever protocol is already being used.

This is related to my comment [here](https://github.com/Expensify/expensify-common/pull/573#discussion_r1319172439), but we can keep this as is for now

File: lib/ExpensiMark.js (hunk 6 above)
Comment (cead22) [reply lib/ExpensiMark.js:494]: > it's out of scope for this PR.

I understand you didn't write this in this PR, but we usually improve code around the code we're editing as we go, and given this isn't super clear, I think we should update it in this PR

> If we find a URL with a leading @ sign, we need look for other valid URL in the rest of the string.

Small correction: If we find a URL with a leading @ sign, we need to look for other valid URLs in the rest of the string.

> If we find another string trailing domain, we apply the auto link rule again and set a flag to avoid doing it

I don't think "string trailing domain" makes sense, how about

If we find another domain in the remainder of the string, we apply the auto link rule again and set a flag to avoid re-doing below

File: __tests__/ExpensiMark-HTML-test.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Antasel) [reply __tests__/ExpensiMark-HTML-test.js:664]: yes. `<a href="https://expensify.com/"...` . is being used

File: lib/ExpensiMark.js (hunk 6 above)
Comment (Antasel) [reply lib/ExpensiMark.js:494]: > I understand you didn't write this in this PR, but we usually improve code around the code we're editing as we go, and given this isn't super clear, I think we should update it in this PR

for the condition `match[1].includes('<pre>')`, we can comment like below.

// We don't want to apply link rule if `match[1]` contains the code block inside the [] of the markdown e.g. `[```example```](https://example.com)`

other conditions's flags are already explained.
What the codes do can be figured out without comments, first one is to concatenate string without applying rule, second one is to apply rule

Comment (Antasel) [reply lib/ExpensiMark.js:494]: > If we find another domain in the remainder of the string, we apply the auto link rule again and set a flag to avoid re-doing below

it's good

Comment (cead22) [reply lib/ExpensiMark.js:494]: Looks good to me 👍

**Comment by Antasel** [comment]
@cead22 Thanks for your review.
Please check new commit

**Review by cead22 — APPROVED** [review]

**Comment by cead22** [comment]
@0xmiroslav wanna give a final review on the latest update before we merge?

**Comment by cead22** [comment]
Thank you! The freeze would apply, but it's over now :)
image
